aeon/dj_pipeline/analysis/block_analysis.py

```diff
@@ -798,3 +786,97 @@ class AnalysisNote(dj.Manual):
     note_type='': varchar(64)
     note: varchar(3000)
     """
+
+
+# ---- Helper Functions ----
+
+
+def get_threshold_associated_pellets(patch_key, start, end):
+    """
+    Retrieve the pellet delivery timestamps associated with each patch threshold update within the specified start-end time.
+    1. Get all patch state update timestamps (DepletionState): let's call these events "A"
+        - Remove all events within 1 second of each other
+        - Remove all events without threshold value (NaN)
+    2. Get all pellet delivery timestamps (DeliverPellet): let's call these events "B"
+        - Find matching beam break timestamps within 500ms after each pellet delivery
+    3. For each event "A", find the nearest event "B" within 100ms before or after the event "A"
+        - These are the pellet delivery events "B" associated with the previous threshold update event "A"
+    4. Shift back the pellet delivery timestamps by 1 to match the pellet delivery with the previous threshold update
+    5. Remove all threshold updates events "A" without a corresponding pellet delivery event "B"
+
+    Args:
+        patch_key (dict): primary key for the patch
+        start (datetime): start timestamp
+        end (datetime): end timestamp
+
+    Returns:
+        pd.DataFrame: DataFrame with the following columns:
+        - threshold_update_timestamp (index)
+        - pellet_timestamp
+        - beam_break_timestamp
+        - offset
+        - rate
+    """
+    chunk_restriction = acquisition.create_chunk_restriction(patch_key["experiment_name"], start, end)
+    # pellet delivery and beam break data
+    delivered_pellet_df = fetch_stream(
+        streams.UndergroundFeederDeliverPellet & patch_key & chunk_restriction
+    )[start:end]
+    beam_break_df = fetch_stream(streams.UndergroundFeederBeamBreak & patch_key & chunk_restriction)[
+        start:end
+    ]
+
+    if delivered_pellet_df.empty or beam_break_df.empty:
+        return acquisition.io_api._empty(
+            ["threshold", "offset", "rate", "pellet_timestamp", "beam_break_timestamp"]
+        )
+
+    # patch threshold data
+    depletion_state_df = fetch_stream(
+        streams.UndergroundFeederDepletionState & patch_key & chunk_restriction
+    )[start:end]
+    # remove NaNs from threshold column
+    depletion_state_df = depletion_state_df.dropna(subset=["threshold"])
+    # remove invalid rows where the time difference is less than 1 second
+    invalid_rows = depletion_state_df.index.to_series().diff().dt.total_seconds() < 1
+    depletion_state_df = depletion_state_df[~invalid_rows]
+
+    # find pellet times with matching beam break times (within 500ms after pellet times)
+    pellet_beam_break_df = (
```

Comment:
Should we remove near-in-time pellet delivery events before doing this `merge_asof`, to reduce the tolerance specified here? Does it make a difference?

I'll look at this briefly

Reply:
Can any of `delivered_pellet_df`, `beam_break_df`, `depletion_state_df` be an empty df? If yes, `pd.merge_asof` will throw an error.

Reply:
That has been checked, the function will return an empty dataframe if `delivered_pellet_df` or `beam_break_df` is empty